File: frontend/src/utils/roles.js
/**
 * Canonical operational roles, mirroring backend/app/auth/roles.py.
 *
 * Roles are stored on `users.role` as a free-form string (the backend has no
 * role enum). This module mirrors the backend's `effective_role` mapping so the
 * frontend gates UI affordances (operator actions, IT log viewer) on the same
 * canonical role the backend enforces. The backend 403 remains authoritative
 * -- this only decides what the UI shows, never what the backend allows.
 */
import { NAVIGATION } from '../data/navigation';

export const ROLE_EMPLOYEE = 'EMPLOYEE';
export const ROLE_OPERATOR = 'OPERATOR';
export const ROLE_REVIEWER = 'REVIEWER';
export const ROLE_IT = 'IT';

//: Legacy role strings mapped to the canonical role they imply. The seeded
//: default account historically used "Verification Officer" and is the
//: all-access Employee account, so it maps to EMPLOYEE.
const ROLE_GROUPS = {
  'Verification Officer': ROLE_EMPLOYEE,
};

const CANONICAL_ROLES = new Set([ROLE_EMPLOYEE, ROLE_OPERATOR, ROLE_REVIEWER, ROLE_IT]);

/**
 * Resolve a stored role string to its canonical role.
 *
 * Matches backend `effective_role`: unrecognized strings pass through unchanged,
 * the empty/missing role defaults to OPERATOR (same as the backend default).
 *
 * @param {string|null|undefined} role Stored `users.role` value, if any.
 * @returns {string} The canonical role.
 */
export function effectiveRole(role) {
  if (!role) {
    return ROLE_OPERATOR;
  }
  if (CANONICAL_ROLES.has(role)) {
    return role;
  }
  return ROLE_GROUPS[role] ?? role;
}

/** Return whether the user holds the operator role. */
export function isOperator(user) {
  return effectiveRole(user?.role) === ROLE_OPERATOR;
}

/** Return whether the user holds the reviewer role. */
export function isReviewer(user) {
  return effectiveRole(user?.role) === ROLE_REVIEWER;
}

/** Return whether the user holds the IT role. */
export function isIt(user) {
  return effectiveRole(user?.role) === ROLE_IT;
}

/**
 * Return whether the user holds the all-access Employee role.
 *
 * Mirrors the backend: EMPLOYEE is a canonical role the seeded
 * "Verification Officer" account maps to, and the backend grants it every
 * operational guard (operator, reviewer and IT actions). Use this at UI gates
 * that should be open to the Employee account (e.g. operator actions, system
 * logs) alongside the specific role predicate.
 *
 * @param {object|null|undefined} user The current user, if any.
 * @returns {boolean} Whether the user holds the all-access Employee role.
 */
export function isEmployee(user) {
  return effectiveRole(user?.role) === ROLE_EMPLOYEE;
}

/**
 * Filter the sidebar navigation for the current user's role.
 *
 * The Employee (all-access) account sees every item so the complete application
 * can be exercised. All other roles only see items whose `roles` list (if any)
 * contains their canonical role. Items without a `roles` array are visible to
 * everyone.
 *
 * @param {object|null|undefined} user The current user, if any.
 * @returns {Array} A deep copy of NAVIGATION with hidden items removed.
 */
export function visibleNavigationFor(user) {
  const effective = effectiveRole(user?.role);

  return NAVIGATION.map((section) => ({
    ...section,
    items: section.items
      .filter((item) => visible(item, effective, user))
      .map((item) =>
        item.children
          ? { ...item, children: item.children.filter((child) => visible(child, effective, user)) }
          : item
      ),
  })).filter((section) => section.items.length > 0);
}

function visible(item, effective, user) {
  if (isEmployee(user)) {
    return true;
  }
  if (!item.roles || item.roles.length === 0) {
    return true;
  }
  return item.roles.includes(effective);
}

File: backend/app/auth/roles.py
"""Operational role definitions and membership helpers.

Roles are stored on ``users.role`` as a free-form string (the existing schema
has no role enum). This module is the single source of truth for the three
operational roles and for mapping legacy role strings onto them so existing
accounts keep working while new role-specific behavior is enforced.

Role hierarchy for authorization:

- EMPLOYEE: the seeded default account (``Verification Officer``) and, by
  design, the all-access role: it is authorized for every operational role
  guard (operator, reviewer and IT actions) so the full application can be
  exercised with one account. This mirrors the product requirement that the
  Employee account can do everything.
- OPERATOR: first-level document/completeness checker. Requests missing
  documents, rejects incomplete applications, submits complete applications
  for processing. Never sees OCR/normalization/technical internals and cannot
  correct extracted fields.
- REVIEWER: second-level verification. Opens applications and documents,
  inspects extracted/normalized values, runs the human-verification workflow,
  approves/corrects/rejects.
- IT: technical/administrative. Reads system logs and operational audit
  information inside the IT settings area. Does not automatically gain
  business approval/review permissions.

Legacy role strings are normalized through :data:`ROLE_GROUPS` so
:func:`effective_role` always returns one of the canonical roles.
"""

from __future__ import annotations

ROLE_EMPLOYEE = "EMPLOYEE"
ROLE_OPERATOR = "OPERATOR"
ROLE_REVIEWER = "REVIEWER"
ROLE_IT = "IT"

#: Canonical roles a user can hold.
ROLES = frozenset({ROLE_EMPLOYEE, ROLE_OPERATOR, ROLE_REVIEWER, ROLE_IT})

#: Legacy role strings mapped to the canonical role they imply. The seeded
#: default account historically used ``Verification Officer`` and is the
#: all-access Employee account, so it maps to EMPLOYEE.
ROLE_GROUPS: dict[str, str] = {
    "Verification Officer": ROLE_EMPLOYEE,
}


def effective_role(role: str | None) -> str:
    """Return the canonical role for a stored role string.

    Args:
        role: Stored ``users.role`` value, if any.

    Returns:
        The canonical role: ``EMPLOYEE``, ``OPERATOR``, ``REVIEWER``, ``IT``,
        or the original string when it is not a recognized legacy value.
    """
    if not role:
        return ROLE_OPERATOR
    if role in ROLES:
        return role
    return ROLE_GROUPS.get(role, role)


def is_employee(user) -> bool:
    """Return whether the user holds the all-access Employee role."""
    return effective_role(getattr(user, "role", None)) == ROLE_EMPLOYEE


def is_operator(user) -> bool:
    """Return whether the user holds the operator role."""
    return effective_role(getattr(user, "role", None)) == ROLE_OPERATOR


def is_reviewer(user) -> bool:
    """Return whether the user holds the reviewer role."""
    return effective_role(getattr(user, "role", None)) == ROLE_REVIEWER


def is_it(user) -> bool:
    """Return whether the user holds the IT role."""
    return effective_role(getattr(user, "role", None)) == ROLE_IT